astartes/samplers/extrapolation/scaffold.py

```diff
@@ -1,5 +1,109 @@
+from collections import defaultdict
+
+import numpy as np
+import rdkit
+from rdkit import Chem
+from rdkit.Chem.Scaffolds import MurckoScaffold
+
 from astartes.samplers import AbstractSampler
+from astartes.utils.warnings import NoMatchingScaffold
 
 
 class Scaffold(AbstractSampler):
-    pass
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        if not isinstance(self.X[0], str) and not isinstance(
+            self.X[0], rdkit.Chem.rdchem.Mol
+        ):
+            msg = "Scaffold class requires input X to be an iterable of SMILES strings"
+            raise TypeError(msg)
+
+    def _sample(self):
+        """Implements the Scaffold sampler to identify clusters via a molecule's Bemis-Murcko scaffold."""
+        scaffold_to_indices = self.scaffold_to_smiles(self.X)
+
+        cluster_indices = np.zeros(len(self.X), dtype=np.int8)
+        # give each cluster an arbitrary ID
+        for cluster_id, (scaffold, indices) in enumerate(scaffold_to_indices.items()):
+            if scaffold == "":
+                warnings.warn(
+                    f"No matching scaffold was found for the {len(indices)} "
+                    f"molecules corresponding to indices {indices}",
+                    NoMatchingScaffold,
+                )
+            for idx in indices:
+                cluster_indices[idx] = cluster_id
+
+        self._samples_clusters = cluster_indices
+
+    def scaffold_to_smiles(self, mols):
+        """
+        Computes scaffold for each smiles string and returns a mapping from scaffolds to sets of smiles.
+
+        Params:
+            mols: A list of smiles strings or RDKit molecules.
+
+        Returns:
+            A dictionary mapping each unique scaffold to all smiles (or smiles indices) which have that scaffold.
+        """
+        scaffolds = defaultdict(set)
+        for i, mol in enumerate(mols):
+            scaffold = self.generate_bemis_murcko_scaffold(mol)
+            scaffolds[scaffold].add(i)
+
+        return scaffolds
+
+    def str_to_mol(self, string, explicit_hydrogens=False):
+        """
+        Converts an InChI or SMILES string to an RDKit molecule.
+
+        Params:
+            string: The InChI or SMILES string.
+            explicit_hydrogens: Whether to treat hydrogens explicitly.
+
+        Returns:
+            An RDKit molecule.
+        """
+        RDKIT_SMILES_PARSER_PARAMS = Chem.SmilesParserParams()
+        if string.startswith("InChI"):
+            mol = Chem.MolFromInchi(string, removeHs=not explicit_hydrogens)
+        else:
+            # Set params here so we don't remove hydrogens with atom mapping
+            RDKIT_SMILES_PARSER_PARAMS.removeHs = not explicit_hydrogens
+            mol = Chem.MolFromSmiles(string, RDKIT_SMILES_PARSER_PARAMS)
+
+        if explicit_hydrogens:
+            return Chem.AddHs(mol)
+        else:
+            return Chem.RemoveHs(mol)
+
+    def generate_bemis_murcko_scaffold(self, mol, include_chirality=False):
```

Comment:
The call to this does not change include_chirality, so it will always be False. If we always want it to be false, we should just have it so. Otherwise we should be retrieving the user configuration from self.get_configs("include_chirality",false)